python-spec/src/somacore/query/fast_csr.py

```diff
@@ -0,0 +1,299 @@
+        self.coo_chunks.append((data.to_numpy(), row_ind, col_ind))
+        _accum_row_length(self.row_length, row_ind)
+
+    def finalize(self) -> _CSRAccumulatorFinalResult:
+        nnz = sum(len(chunk[0]) for chunk in self.coo_chunks)
+        index_dtype = _select_dtype(nnz)
+        if nnz == 0:
+            # no way to infer matrix dtype, so use default and return empty matrix
+            empty = sparse.csr_matrix((0, 0))
+            return _CSRAccumulatorFinalResult(
+                data=empty.data,
+                indptr=empty.indptr,
+                indices=empty.indices,
+                shape=(0, 0),
+            )
+
+        # cumsum row lengths to get indptr
+        indptr = np.empty((self.shape[0] + 1,), dtype=index_dtype)
+        indptr[0:1] = 0
+        np.cumsum(self.row_length, out=indptr[1:])
+
+        # Parallel copy of data and column indices
+        indices = np.empty((nnz,), dtype=index_dtype)
+        data = np.empty((nnz,), dtype=self.coo_chunks[0][0].dtype)
+
+        row_rng_mask_bits = 18
+        n_jobs = (self.shape[0] >> row_rng_mask_bits) + 1
+        chunk_list = numba.typed.List(self.coo_chunks)
+        futures.wait(
+            [
+                self.pool.submit(
+                    _copy_chunklist_range,
+                    chunk_list,
+                    data,
+                    indices,
+                    indptr,
+                    row_rng_mask_bits,
+                    job,
+                )
+                for job in range(n_jobs)
+            ]
+        )
+        _finalize_indptr(indptr)
+        return _CSRAccumulatorFinalResult(
+            data=data, indptr=indptr, indices=indices, shape=self.shape
+        )
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _accum_row_length(
+    row_length: npt.NDArray[np.int64], row_ind: npt.NDArray[np.int64]
+) -> None:
+    for rind in row_ind:
+        row_length[rind] += 1
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _copy_chunk_range(
+    data_chunk: npt.NDArray[np.number],
+    row_ind_chunk: npt.NDArray[np.signedinteger],
+    col_ind_chunk: npt.NDArray[np.signedinteger],
+    data: npt.NDArray[np.number],
+    indices: npt.NDArray[np.signedinteger],
+    indptr: npt.NDArray[np.signedinteger],
+    row_rng_mask: int,
+    row_rng_val: int,
+):
+    for n in range(len(data_chunk)):
+        row = row_ind_chunk[n]
+        if (row & row_rng_mask) != row_rng_val:
+            continue
+        ptr = indptr[row]
+        indices[ptr] = col_ind_chunk[n]
+        data[ptr] = data_chunk[n]
+        indptr[row] += 1
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _copy_chunklist_range(
+    chunk_list: numba.typed.List,
+    data: npt.NDArray[np.number],
+    indices: npt.NDArray[np.signedinteger],
+    indptr: npt.NDArray[np.signedinteger],
+    row_rng_mask_bits: int,
+    job: int,
+):
+    row_rng_mask = (2**64 - 1) >> row_rng_mask_bits << row_rng_mask_bits
```

Comment:
do we need to assert `row_rng_mask_bits` is less than 64? here it's 18 (currently) but if any well-meaning soul every changes that value down there without looking up here, i suspect they'll be doing a bit of debugging to figure out what's going wrong ...